# MOVIE RECOMMENDATION SYSTEM

Use pandas to read dataset.
Use numpy to create arrays

In [4]:
import pandas as pd  
import numpy as np 

Dataset: https://www.kaggle.com/tmdb/tmdb-movie-metadata
There are 2 datasets: credits & movies

In [22]:
credits = pd.read_csv("tmdb_5000_credits.csv")

In [23]:
credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4804 entries, 0 to 4803
Data columns (total 56 columns):
movie_id       4804 non-null int64
title          4804 non-null object
cast           4804 non-null object
crew           4800 non-null object
Unnamed: 4     1 non-null object
Unnamed: 5     1 non-null object
Unnamed: 6     1 non-null object
Unnamed: 7     1 non-null object
Unnamed: 8     1 non-null object
Unnamed: 9     1 non-null object
Unnamed: 10    1 non-null object
Unnamed: 11    1 non-null object
Unnamed: 12    1 non-null object
Unnamed: 13    1 non-null object
Unnamed: 14    1 non-null object
Unnamed: 15    1 non-null object
Unnamed: 16    1 non-null object
Unnamed: 17    1 non-null object
Unnamed: 18    1 non-null object
Unnamed: 19    1 non-null object
Unnamed: 20    1 non-null object
Unnamed: 21    1 non-null object
Unnamed: 22    1 non-null object
Unnamed: 23    0 non-null float64
Unnamed: 24    0 non-null float64
Unnamed: 25    0 non-null float64
Unnamed: 26    0 non-

In [24]:
credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew Unnamed: 4 Unnamed: 5  \
0  [{"credit_id": "52fe48009251416c750aca23", "de...        NaN        NaN   
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...        NaN        NaN   
2  [{"credit_id": "54805967c3a36829b5002c41", "de...        NaN        NaN   
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...        NaN        NaN   
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...        NaN        NaN   

  Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 46 Unnamed: 47  \
0        NaN        NaN        NaN        NaN  ...         NaN         NaN   
1        NaN        NaN        NaN        NaN  ...         NaN         NaN   
2        NaN        NaN        NaN        NaN  ...         NaN         NaN   
3        NaN        NaN        NaN        NaN  ...         NaN         NaN   
4        NaN        NaN        NaN        NaN  ...         NaN         NaN   

  Unnamed: 48 Unnamed: 49 Unnamed: 50 Unnamed: 51 Unnamed: 52 Unnamed: 53  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 54  id  
0         NaN NaN  
1         NaN NaN  
2         NaN NaN  
3         NaN NaN  
4         NaN NaN  

[5 rows x 56 columns]

In [25]:
print(credits.columns)

Index(['movie_id', 'title', 'cast', 'crew', 'Unnamed: 4', 'Unnamed: 5',
       'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10',
       'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14',
       'Unnamed: 15', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20', 'Unnamed: 21', 'Unnamed: 22',
       'Unnamed: 23', 'Unnamed: 24', 'Unnamed: 25', 'Unnamed: 26',
       'Unnamed: 27', 'Unnamed: 28', 'Unnamed: 29', 'Unnamed: 30',
       'Unnamed: 31', 'Unnamed: 32', 'Unnamed: 33', 'Unnamed: 34',
       'Unnamed: 35', 'Unnamed: 36', 'Unnamed: 37', 'Unnamed: 38',
       'Unnamed: 39', 'Unnamed: 40', 'Unnamed: 41', 'Unnamed: 42',
       'Unnamed: 43', 'Unnamed: 44', 'Unnamed: 45', 'Unnamed: 46',
       'Unnamed: 47', 'Unnamed: 48', 'Unnamed: 49', 'Unnamed: 50',
       'Unnamed: 51', 'Unnamed: 52', 'Unnamed: 53', 'Unnamed: 54', 'id'],
      dtype='object')


# Data cleaning

In [9]:
credits_clean = credits.drop(credits.columns[4:],axis=1)

In [10]:
credits_clean.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [11]:
movies = pd.read_csv("tmdb_5000_movies.csv")

In [12]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
budget                  4803 non-null int64
genres                  4803 non-null object
homepage                1712 non-null object
id                      4803 non-null int64
keywords                4803 non-null object
original_language       4803 non-null object
original_title          4803 non-null object
overview                4800 non-null object
popularity              4803 non-null float64
production_companies    4803 non-null object
production_countries    4803 non-null object
release_date            4802 non-null object
revenue                 4803 non-null int64
runtime                 4801 non-null float64
spoken_languages        4803 non-null object
status                  4803 non-null object
tagline                 3959 non-null object
title                   4803 non-null object
vote_average            4803 non-null float64
vote_count              4803 non-null 

In [13]:
print(movies.columns)

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')


In [14]:
movies.head()


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...     12/10/09  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...      5/19/07   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...     10/26/15   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...      7/16/12  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...       3/7/12   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [15]:
credits_renamed = credits_clean.rename(index=str, columns={"movie_id": "id"})
credits_renamed

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4799    9367                               El Mariachi   
4800   72766                                 Newlyweds   
4801  231617                 Signed, Sealed, Delivered   
4802  126186                          Shanghai Calling   
4803   25975                         My Date with Drew   

                                                   cast  \
0     [{"cast_id": 242, "character": "Jake Sully", "...   
1     [{"cast_id": 4, "character": "Captain Jack Spa...   
2     [{"cast_id": 1, "character": "James Bond", "cr...   
3     [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4     [{"cast_id": 5, "character": "John Carter", "c...   
...                                                 ...   
4799  [{"cast_id": 1, "character": "El Mariachi", "c...   
4800  [{"cast_id": 1, "character": "Buzzy", "credit_...   
4801  [{"cast_id": 8, "character": "Oliver O\u2019To...   
4802  [{"cast_id": 3, "character": "Sam", "credit_id...   
4803  [{"cast_id": 3, "character": "Herself", "credi...   

                                                   crew  
0     [{"credit_id": "52fe48009251416c750aca23", "de...  
1     [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2     [{"credit_id": "54805967c3a36829b5002c41", "de...  
3     [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4     [{"credit_id": "52fe479ac3a36847f813eaa3", "de...  
...                                                 ...  
4799  [{"credit_id": "52fe44eec3a36847f80b280b", "de...  
4800  [{"credit_id": "52fe487dc3a368484e0fb013", "de...  
4801  [{"credit_id": "52fe4df3c3a36847f8275ecf", "de...  
4802  [{"credit_id": "52fe4ad9c3a368484e16a36b", "de...  
4803  [{"credit_id": "58ce021b9251415a390165d9", "de...  

[4804 rows x 4 columns]

In [16]:
movies_merge = movies.merge(credits_renamed, on='id')
movies_merge.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                    title_x vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                    title_y  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne /

In [17]:
movies_clean = movies_merge.drop(columns=['homepage', 'title_x', 'title_y', 'status','production_countries'])
movies_clean.head()

budget                                             genres      id  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies release_date     revenue  \
0  [{"name": "Ingenious Film Partners", "id": 289...     12/10/09  2787965087   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...      5/19/07   961000000   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...     10/26/15   880674609   
3  [{"name": "Legendary Pictures", "id": 923}, {"...      7/16/12  1084939099   
4        [{"name": "Walt Disney Pictures", "id": 2}]       3/7/12   284139100   

   runtime                                   spoken_languages  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...   
1    169.0           [{"iso_639_1": "en", "name": "English"}]   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...   
3    165.0           [{"iso_639_1": "en", "name": "English"}]   
4    132.0           [{"iso_639_1": "en", "name": "English"}]   

                                          tagline  vote_average  vote_count  \
0                     Enter the World of Pandora.           7.2       11800   
1  At the end of the world, the adventure begins.           6.9        4500   
2                           A Plan No One Escapes           6.3        4466   
3                                 The Legend Ends           7.6        9106   
4            Lost in our world, found in another.           6.1        2124   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [18]:
print(movies_clean.columns)

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'release_date', 'revenue', 'runtime', 'spoken_languages', 'tagline',
       'vote_average', 'vote_count', 'cast', 'crew'],
      dtype='object')


# Content Based Recommendation system

recommendations based on the movie’s plot summaries given in the overview column. 
So if user gives a movie title, goal is to recommend movies that share similar plot summaries.

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [21]:
movies_clean.head()['overview']

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

For recommendation engine we need to create a vector of matrix because recommendation system is based on pairwise similarity so we represent every movie in terms of vectors

"Overview" is a string column and model cannot understand a sentence
For this we use NLP concept "tfidf"

Tfidf technique helps us to covert sentences into document matrix

min_df=3: ignore the terms that have frequency less than threshold value=3
max_features=none:

strip_accent='unicode': Remove accents and perform other character normalization;‘ascii’ is a fast method that only works on characters that have an direct ASCII mapping. ‘unicode’ is a slightly slower method that works on any characters.

analyzer='word': Whether the feature should be made of word or character n-grams

token_pattern=r'\w{1,}': remove unnecessary punctuations

ngram_range=(1, 3): lower and upper boundary of the range of n-values- take combination of 1 to 3 different kind of words

stop_words = 'english': remove all unnecessary characters like "the,at,of,is,etc." since these chars exists multiple number of times

In [27]:
tfv = TfidfVectorizer(min_df=3,  max_features=None, 
            strip_accents='unicode', analyzer='word',token_pattern=r'\w{1,}',
            ngram_range=(1, 3),
            stop_words = 'english')

Filling NaNs with empty string

In [28]:
movies_clean['overview'] = movies_clean['overview'].fillna('')

Fitting the TF-IDF on the 'overview' text

Fit_transform converts it into a sparse matrix. Sparse matrix has many 0 and less non-zero values

In [29]:
tfv_matrix = tfv.fit_transform(movies_clean['overview'])

In [30]:
tfv_matrix

<4803x10417 sparse matrix of type '<class 'numpy.float64'>'
	with 127220 stored elements in Compressed Sparse Row format>

In [31]:
tfv_matrix.shape

(4803, 10417)

So we have 4803 records and 10417 different features. Number of features depend upon ngram. Combination of different words become our columns

In [32]:
from sklearn.metrics.pairwise import sigmoid_kernel

we use sigmoid to transform our input between 0 & 1. This Sigmoid is also used in logistic regression. Purpose is to check how overview1 is related to overview2

To Compute the sigmoid kernel use tfv_matrix twice cz only then the combination of matrix can be done. Everything will be based on the vectors

In [33]:
sig = sigmoid_kernel(tfv_matrix, tfv_matrix)

In [37]:
sig[0]

array([0.76163447, 0.76159416, 0.76159416, ..., 0.76159416, 0.76159416,
       0.76159416])

s = summary = overview

eg: (s1,s1)= 0.76163447 (s1,s2)= 0.76159416 and so on

Perform Reverse mapping of indices and movie titles.
Creating indices series with title and index in movies dataset

In [39]:
indices = pd.Series(movies_clean.index, index=movies_clean['original_title']).drop_duplicates()


In [40]:
indices

original_title
Avatar                                         0
Pirates of the Caribbean: At World's End       1
Spectre                                        2
The Dark Knight Rises                          3
John Carter                                    4
                                            ... 
El Mariachi                                 4798
Newlyweds                                   4799
Signed, Sealed, Delivered                   4800
Shanghai Calling                            4801
My Date with Drew                           4802
Length: 4803, dtype: int64

In [ ]:
Instead of performing individual steps multiple times; we can a write function 

In [41]:
indices['John Carter']

4

In [42]:
sig[4]

array([0.76159416, 0.76159579, 0.76159442, ..., 0.76159463, 0.76159416,
       0.7615949 ])

In [44]:
list(enumerate(sig[4]))

[(0, 0.7615941559557649),
 (1, 0.7615957851727588),
 (2, 0.7615944213582981),
 (3, 0.7615950457737669),
 (4, 0.7616344692549826),
 (5, 0.7615943590937618),
 (6, 0.7615953553357809),
 (7, 0.7615961583389188),
 (8, 0.761594434644127),
 (9, 0.7615955376945788),
 (10, 0.7615950187584506),
 (11, 0.7615941559557649),
 (12, 0.7615955245177014),
 (13, 0.7615941559557649),
 (14, 0.7615945776557239),
 (15, 0.7615941559557649),
 (16, 0.7615982150807789),
 (17, 0.761595905734489),
 (18, 0.7615948563013252),
 (19, 0.7615954601866739),
 (20, 0.761594742874267),
 (21, 0.7615959017864065),
 (22, 0.7615941559557649),
 (23, 0.7615944798687346),
 (24, 0.7615949165564755),
 (25, 0.7615941559557649),
 (26, 0.761596672208609),
 (27, 0.761596811586298),
 (28, 0.7615956924959928),
 (29, 0.7615950426593034),
 (30, 0.761594370244919),
 (31, 0.7615973778675746),
 (32, 0.7615952217232149),
 (33, 0.761594480253419),
 (34, 0.7615941559557649),
 (35, 0.7615948292488448),
 (36, 0.7615961922750802),
 (37, 0.7615944401

In [45]:
sorted(list(enumerate(sig[indices['John Carter']])), key=lambda x: x[1], reverse=True)

[(4, 0.7616344692549826),
 (1254, 0.7616028441227217),
 (4161, 0.7616025058398342),
 (2932, 0.7615997702387518),
 (3349, 0.7615994494540442),
 (1307, 0.7615993499452608),
 (91, 0.7615993393002068),
 (345, 0.7615991005452573),
 (3068, 0.7615989528792844),
 (4274, 0.7615984483964565),
 (581, 0.7615983846894889),
 (16, 0.7615982150807789),
 (1300, 0.7615981269760195),
 (1735, 0.7615980797565637),
 (2968, 0.7615978858188849),
 (2755, 0.7615978558146024),
 (1891, 0.7615978409791568),
 (2300, 0.7615977881759383),
 (1057, 0.761597768927918),
 (2925, 0.7615976750553425),
 (3270, 0.7615976743123604),
 (773, 0.7615976367651353),
 (2998, 0.7615976200564404),
 (694, 0.7615976012840266),
 (3373, 0.761597581185098),
 (2417, 0.761597528949232),
 (46, 0.7615974954501349),
 (2053, 0.7615974142279796),
 (322, 0.7615973929787131),
 (2966, 0.761597388185314),
 (31, 0.7615973778675746),
 (1974, 0.7615973685609891),
 (755, 0.7615972915924529),
 (2265, 0.7615972649713753),
 (2900, 0.7615972599206641),
 (3914

In [48]:
def give_rec(title, sig=sig):
    idx = indices[title] # Get the index corresponding to original_title
    sig_scores = list(enumerate(sig[idx])) # Get the pairwsie similarity scores 
    sig_scores = sorted(sig_scores, key=lambda x: x[1], reverse=True) # Sort the movies 
    sig_scores = sig_scores[1:11] # Scores of the 10 most similar movies
    movie_indices = [i[0] for i in sig_scores]     # Movie indices
    return movies_clean['original_title'].iloc[movie_indices]  # Top 10 most similar movies

Content-based recommendation system 


In [50]:
give_rec('Newlyweds')

616                       Ted 2
2689         Our Family Wedding
869          You, Me and Dupree
3969           Something Wicked
1576                 Bride Wars
2290               Just Married
1032      America's Sweethearts
3145                      Amour
2531     Why Did I Get Married?
504     The Secret Life of Pets
Name: original_title, dtype: object